pydantic/_internal/_generate_schema.py

```diff
@@ -1126,9 +1122,16 @@ def _apply_annotations(
             prepare = getattr(annotation, '__prepare_pydantic_annotations__', None)
             if prepare is not None:
                 previous = annotations[:idx]
-                remaining = annotations[idx:]
-                remaining = list(prepare(source_type, tuple(remaining)))
+                remaining = annotations[idx + 1 :]
+                new_source_type, remaining = prepare(source_type, tuple(remaining))
                 annotations = previous + remaining
+                if new_source_type is not source_type:
+                    return self._apply_annotations(
+                        get_inner_schema,
+                        new_source_type,
+                        annotations,
+                        check_prepare_on_source=check_prepare_on_source,
+                    )
```

Comment:
This is the only part that's a bit wonky about this PR: if something replaces the source type we "start over" going from right to left again. My main fear is that it might be a bit difficult to explain, but this is a _very_ advanced use case so maybe that's okay.